File: Labs/Lab3/src/api/orchidService.js
import axios from './axios';

export const orchidService = {
  // GET - Lấy tất cả orchids
  getAllOrchids: async () => {
    try {
      const response = await axios.get('/orchids');
      return response.data;
    } catch (error) {
      throw error;
    }
  },

  // GET - Lấy orchid theo ID
  getOrchidById: async (id) => {
    try {
      const response = await axios.get(`/orchids/${id}`);
      return response.data;
    } catch (error) {
      throw error;
    }
  },

  // POST - Tạo orchid mới (FIX ID)
  createOrchid: async (orchidData) => {
    try {
      // Lấy tất cả orchids để tìm ID lớn nhất
      const allOrchids = await axios.get('/orchids');
      const maxId = allOrchids.data.reduce((max, orchid) => {
        const id = parseInt(orchid.id);
        return id > max ? id : max;
      }, 0);
      
      // Tạo orchid mới với ID tuần tự
      const newOrchid = {
        ...orchidData,
        id: String(maxId + 1)
      };
      
      const response = await axios.post('/orchids', newOrchid);
      return response.data;
    } catch (error) {
      throw error;
    }
  },

  // PUT - Cập nhật orchid
  updateOrchid: async (id, orchidData) => {
    try {
      const response = await axios.put(`/orchids/${id}`, {
        ...orchidData,
        id: id  // Đảm bảo ID không thay đổi
      });
      return response.data;
    } catch (error) {
      throw error;
    }
  },

  // DELETE - Xóa orchid
  deleteOrchid: async (id) => {
    try {
      const response = await axios.delete(`/orchids/${id}`);
      return response.data;
    } catch (error) {
      throw error;
    }
  },

  // GET - Search orchids
  searchOrchids: async (query) => {
    try {
      const response = await axios.get(`/orchids?q=${query}`);
      return response.data;
    } catch (error) {
      throw error;
    }
  },

  // GET - Filter by category
  filterByCategory: async (category) => {
    try {
      const response = await axios.get(`/orchids?category=${category}`);
      return response.data;
    } catch (error) {
      throw error;
    }
  }
};

File: Labs/Lab5/src/main/java/nguyenphamhoangminh/lab5/services/OrchidService.java
package nguyenphamhoangminh.lab5.services;

import lombok.RequiredArgsConstructor;
import nguyenphamhoangminh.lab5.exception.BadRequestException;
import nguyenphamhoangminh.lab5.exception.ResourceNotFoundException;
import nguyenphamhoangminh.lab5.pojos.Category;
import nguyenphamhoangminh.lab5.pojos.Orchid;
import nguyenphamhoangminh.lab5.repositories.CategoryRepository;
import nguyenphamhoangminh.lab5.repositories.OrchidRepository;
import org.springframework.stereotype.Service;
import java.util.List;

@Service
@RequiredArgsConstructor
public class OrchidService {

    private final OrchidRepository orchidRepository;
    private final CategoryRepository categoryRepository;

    // Lấy tất cả orchids
    public List<Orchid> getAllOrchids() {
        return orchidRepository.findAll();
    }

    // Lấy orchid theo ID
    public Orchid getOrchidById(Long id) {
        return orchidRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Orchid", "id", id));
    }

    // Lấy orchids theo category
    public List<Orchid> getOrchidsByCategory(Long categoryId) {
        // Kiểm tra category có tồn tại không
        if (!categoryRepository.existsById(categoryId)) {
            throw new ResourceNotFoundException("Category", "id", categoryId);
        }
        return orchidRepository.findByCategoryId(categoryId);
    }

    // Tìm kiếm orchids theo tên
    public List<Orchid> searchOrchidsByName(String name) {
        return orchidRepository.findByOrchidNameContainingIgnoreCase(name);
    }

    // Tạo mới orchid
    public Orchid createOrchid(Orchid orchid) {
        // Validate
        validateOrchid(orchid);

        // Kiểm tra category có tồn tại không
        Category category = categoryRepository.findById(orchid.getCategory().getId())
                .orElseThrow(() -> new ResourceNotFoundException("Category", "id", orchid.getCategory().getId()));

        orchid.setCategory(category);
        return orchidRepository.save(orchid);
    }

    // Cập nhật orchid
    public Orchid updateOrchid(Long id, Orchid orchidDetails) {
        Orchid orchid = getOrchidById(id);

        // Validate
        validateOrchid(orchidDetails);

        // Kiểm tra category có tồn tại không
        Category category = categoryRepository.findById(orchidDetails.getCategory().getId())
                .orElseThrow(() -> new ResourceNotFoundException("Category", "id", orchidDetails.getCategory().getId()));

        orchid.setOrchidName(orchidDetails.getOrchidName());
        orchid.setPrice(orchidDetails.getPrice());
        orchid.setDescription(orchidDetails.getDescription());
        orchid.setImage(orchidDetails.getImage());
        orchid.setCategory(category);

        return orchidRepository.save(orchid);
    }

    // Xóa orchid
    public void deleteOrchid(Long id) {
        Orchid orchid = getOrchidById(id);
        orchidRepository.delete(orchid);
    }

    // Validate orchid data
    private void validateOrchid(Orchid orchid) {
        if (orchid.getOrchidName() == null || orchid.getOrchidName().trim().isEmpty()) {
            throw new BadRequestException("Orchid name cannot be empty");
        }

        if (orchid.getPrice() == null || orchid.getPrice() <= 0) {
            throw new BadRequestException("Price must be greater than 0");
        }

        if (orchid.getCategory() == null || orchid.getCategory().getId() == null) {
            throw new BadRequestException("Category is required");
        }
    }
}